charms/istio-pilot/src/charm.py

```diff
@@ -286,64 +294,106 @@ def handle_ingress_auth(self, event):
             )
             return
 
-        t = self.env.get_template('auth_filter.yaml.j2')
-        auth_filters = ''.join(
+        t = self.env.get_template("auth_filter.yaml.j2")
+        auth_filters = "".join(
             t.render(
                 namespace=self.model.name,
                 app_name=self.app.name,
                 **{
-                    'request_headers': yaml.safe_dump(
-                        [{'exact': h} for h in r.get('allowed-request-headers', [])],
+                    "request_headers": yaml.safe_dump(
+                        [{"exact": h} for h in r.get("allowed-request-headers", [])],
                         default_flow_style=True,
                     ),
-                    'response_headers': yaml.safe_dump(
-                        [{'exact': h} for h in r.get('allowed-response-headers', [])],
+                    "response_headers": yaml.safe_dump(
+                        [{"exact": h} for h in r.get("allowed-response-headers", [])],
                         default_flow_style=True,
                     ),
-                    'port': r['port'],
-                    'service': r['service'],
+                    "port": r["port"],
+                    "service": r["service"],
                 },
             )
             for r in auth_routes
         )
 
         self._resource_handler.delete_existing_resources(
             self._resource_handler.get_custom_resource_class_from_filename(
-                filename='auth_filter.yaml.j2'
+                filename="auth_filter.yaml.j2"
             ),
             namespace=self.model.name,
         )
         self._resource_handler.apply_manifest(auth_filters, namespace=self.model.name)
 
-    @property
-    def _gateway_address(self):
-        """Look up the load balancer address for the ingress gateway.
-        If the gateway isn't available or doesn't have a load balancer address yet,
-        returns None.
-        """
-        # FIXME: service name is hardcoded
+    def _is_gateway_service_up(self):
+        """Returns True if the ingress gateway service is up, else False."""
+        svc = self._get_gateway_service()
+
+        if svc.spec.type == "NodePort":
+            # TODO: do we need to interrogate this further for status?
+            return True
+        if _get_gateway_address_from_svc(svc) is not None:
```

Comment:
`_get_gateway_address_from_svc` returns address from service .... If I am not wrong we never use the address so I am just curious if its finding the address of the services worth the hustle. If not we can just remove the whole `_get_address_from_loadbalancer` and return true if the service is there. Or is there a chance the service is up without address ? if so dont we also need to check some more info about the nodeport service ? (e.g. if port was really assigned).

Reply:
You're right, we never use the address as an address.  It is a surrogate for whether the LoadBalancer is up and functioning like you mention.  That the service exists is not enough (for example, think of a LoadBalancer service in microk8s when we haven't deployed metallb - it is a service, but it does not have an ip).  It doesn't feel like the best approach, but I don't think we've come up with something better.

For the NodePort, I agree that it would be better to interrogate it further, but I didn't see anything more available.  It doesn't have a status or other attributes that indicated anything.  I don't think it can be up without a port, but I'm not sure